networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/PeerManager.java

```diff
@@ -83,36 +84,34 @@ public void unsubscribeConnect(final long subscriptionId) {
 
   public SafeFuture<?> connect(final Multiaddr peer, final Network network) {
     STDOUT.log(Level.DEBUG, "Connecting to " + peer);
-    return SafeFuture.of(network.connect(peer))
-        .whenComplete(
-            (conn, throwable) -> {
-              if (throwable != null) {
-                STDOUT.log(
-                    Level.DEBUG,
-                    "Connection to " + peer + " failed. Will retry shortly: " + throwable);
-                ignoreFuture(
-                    scheduler.schedule(
-                        () -> connect(peer, network).reportExceptions(),
-                        RECONNECT_TIMEOUT,
-                        TimeUnit.MILLISECONDS));
-              } else {
-                STDOUT.log(
-                    Level.DEBUG,
-                    "Connection to peer: "
-                        + conn.secureSession().getRemoteId()
-                        + " was successful");
-                SafeFuture.of(conn.closeFuture())
-                    .finish(
-                        () -> {
-                          LOG.debug("Connection to {} closed. Will retry shortly", peer);
-                          ignoreFuture(
-                              scheduler.schedule(
-                                  () -> connect(peer, network).reportExceptions(),
-                                  RECONNECT_TIMEOUT,
-                                  TimeUnit.MILLISECONDS));
-                        });
-              }
-            });
+    final SafeFuture<Connection> initialConnectionFuture = SafeFuture.of(network.connect(peer));
+    final SafeFuture<?> retryLoop =
+        initialConnectionFuture
+            .thenCompose(
+                conn -> {
+                  LOG.debug(
+                      "Connection to peer {} was successful", conn.secureSession().getRemoteId());
+                  return SafeFuture.of(conn.closeFuture());
+                })
+            .exceptionally(
+                (err) -> {
+                  LOG.debug("Connection to {} failed: {}", peer, err);
+                  return null;
+                })
+            .thenCompose(
+                (res) -> {
+                  LOG.debug(
+                      "Connection to {} was closed. Will retry in {} sec",
+                      peer,
+                      RECONNECT_TIMEOUT.toSeconds());
+
+                  return asynRunner.runAfterDelay(
+                      () -> connect(peer, network),
+                      RECONNECT_TIMEOUT.toMillis(),
+                      TimeUnit.MILLISECONDS);
+                });
+    ignoreFuture(retryLoop);
```

Comment:
Given that this shouldn't ever complete exceptionally, could we use `reportExceptions()` on the end instead of `ignoreFuture`?  That way if for some reason an exception does bubble up because of a bug somewhere it will get logged instead of silently ignored.

Reply:
👍updated